File: 3414.java
//* @author:
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Scanner;

public class Main{
 private static final String[] status = new String[] { "", "FILL(1)",
    "FILL(2)", "DROP(1)", "DROP(2)", "POUR(1,2)", "POUR(2,1)" };
 private static boolean[][] visted = new boolean[101][101];
 private static int a;
 private static int b;
 private static int c;

 public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    for (int i = 0; i < 101; ++i) {
        for (int j = 0; j < 101; ++j)
            visted[i][j] = false;
    }
    a = sc.nextInt();
    b = sc.nextInt();
    c = sc.nextInt();
    bfs();
  }

  static void bfs() {
    boolean is = false;
    LinkedList< Node> queue = new LinkedList< Node>();
    Node p = new Node();
    p.aa = 0;
    p.bb = 0;
    visted[p.aa][p.bb] = true;
    queue.addLast(p);
    // visted[p.aa][p.bb] = true;
    // p = new Node();
    // p.aa = 0;
    // p.bb = b;
    // p.status.add(2);
    // visted[p.aa][p.bb] = true;
    // queue.addLast(p);
    while (!queue.isEmpty()) {
     Node front = queue.getFirst();
     queue.remove();
     int r = -1;
     for (int i = 1; i <= 6; ++i) {
      int aa = front.aa;
      int bb = front.bb;
      int flag = i;
      switch (flag) {
       case 1:
        aa = a;
        break;
       case 2:
        bb = b;
        break;
       case 3:
        aa = 0;
        break;
       case 4:
        bb = 0;
        break;
       case 5:
        r = b - bb;
        bb = bb + (r <= aa ? r : aa);
        aa = (r <= aa ? (aa - r) : 0);
        break;
       case 6:
        r = a - aa;
        aa = aa + (r <= bb ? r : bb);
        bb = (r <= bb ? (bb - r) : 0);
        break;
       default:
        break;
       }
       if (aa == c || bb == c) {
        int size = front.status.size();
        System.out.println(size + 1);
        for (int j = 0; j < size; ++j)
             System.out.println(status[front.status.get(j)]);
        System.out.println(status[flag]);
        is = true;
        return;
        }
        if (!visted[aa][bb]) {
         Node tmp = new Node();
         tmp.aa = aa;
         tmp.bb = bb;
         int size = front.status.size();
         for (int j = 0; j < size; ++j)
            tmp.status.add(front.status.get(j));
         tmp.status.add(flag);
         visted[aa][bb] = true;
         queue.addLast(tmp);
         }
     }
    }
    if (!is)
      System.out.println("impossible");
 }

 private static class Node {
    private int aa;
    private int bb;
    private ArrayList< Integer> status = new ArrayList< Integer>();
 }
}

File: 3414.cc
#include <iostream>
#include <queue>
using namespace std;
#define MAXV 110

typedef struct Node {
  int aa, bb, type, cout;
  int i;
  int pa, pb, ca, cb;
} VNode;

VNode map[MAXV][MAXV], v, temp;
bool dis[MAXV][MAXV];

void write(int first, int last) {
  int ft, lt;
  printf("%d\n", map[first][last].cout);

  map[first][last].ca = -1, map[first][last].cb = -1;

  while (map[first][last].pa != -1 && map[first][last].pb != -1) {
    ft = first, lt = last;
    v = map[map[first][last].pa][map[first][last].pb];
    map[map[first][last].pa][map[first][last].pb].ca = first,
    map[map[first][last].pa][map[first][last].pb].cb = last;
    first = v.aa, last = v.bb;
  }
  while (map[first][last].ca != -1 && map[first][last].cb != -1) {
    ft = first, lt = last;
    switch (map[map[first][last].ca][map[first][last].cb].type) {
      case 1:
        printf("FILL(%d)\n", map[map[first][last].ca][map[first][last].cb].i);
        break;
      case 2:
        printf("POUR(%d,%d)\n",
               map[map[first][last].ca][map[first][last].cb].i ^ 3,
               map[map[first][last].ca][map[first][last].cb].i);
        break;
      case 3:
        printf("DROP(%d)\n", map[map[first][last].ca][map[first][last].cb].i);
        break;
    }

    first = map[ft][lt].ca, last = map[ft][lt].cb;
  }
}

int bfs(int a, int b, int c) {
  queue<VNode> q;
  memset(dis, false, sizeof(dis));
  memset(map, 0, sizeof(map));

  dis[0][0] = true;
  map[0][0].pa = -1, map[0][0].pb = -1;
  q.push(map[0][0]);

  while (!q.empty()) {
    temp = v = q.front();
    q.pop();

    v.aa = a;
    if (!dis[v.aa][v.bb]) {
      dis[v.aa][v.bb] = true;
      v.cout = temp.cout + 1;
      v.pa = temp.aa, v.pb = temp.bb;
      v.type = 1;
      v.i = 1;
      map[v.aa][v.bb] = v;
      if (v.aa == c || v.bb == c) {
        write(v.aa, v.bb);
        return 0;
      }
      q.push(v);
    }
    v = temp;
    v.bb = b;
    if (!dis[v.aa][v.bb]) {
      dis[v.aa][v.bb] = true;
      v.cout = temp.cout + 1;
      v.pa = temp.aa, v.pb = temp.bb;
      v.type = 1;
      v.i = 2;
      map[v.aa][v.bb] = v;
      if (v.aa == c || v.bb == c) {
        write(v.aa, v.bb);
        return 0;
      }
      q.push(v);
    }
    v = temp;
    v.aa = v.bb + v.aa;
    if (v.aa > a) {
      v.bb = v.aa - a;
      v.aa = a;
    } else
      v.bb = 0;
    if (!dis[v.aa][v.bb]) {
      dis[v.aa][v.bb] = true;
      v.cout = temp.cout + 1;
      v.pa = temp.aa, v.pb = temp.bb;
      v.type = 2;
      v.i = 1;
      map[v.aa][v.bb] = v;
      if (v.aa == c || v.bb == c) {
        write(v.aa, v.bb);
        return 0;
      }
      q.push(v);
    }
    v = temp;
    v.bb = v.bb + v.aa;
    if (v.bb > b) {
      v.aa = v.bb - b;
      v.bb = b;
    } else
      v.aa = 0;
    if (!dis[v.aa][v.bb]) {
      dis[v.aa][v.bb] = true;
      v.cout = temp.cout + 1;
      v.pa = temp.aa, v.pb = temp.bb;
      v.type = 2;
      v.i = 2;
      map[v.aa][v.bb] = v;
      if (v.aa == c || v.bb == c) {
        write(v.aa, v.bb);
        return 0;
      }
      q.push(v);
    }
    v = temp;
    v.aa = 0;
    if (!dis[v.aa][v.bb]) {
      dis[v.aa][v.bb] = true;
      v.cout = temp.cout + 1;
      v.pa = temp.aa, v.pb = temp.bb;
      v.type = 3;
      v.i = 1;
      map[v.aa][v.bb] = v;
      if (v.aa == c || v.bb == c) {
        write(v.aa, v.bb);
        return 0;
      }
      q.push(v);
    }
    v = temp;
    v.bb = 0;
    if (!dis[v.aa][v.bb]) {
      dis[v.aa][v.bb] = true;
      v.cout = temp.cout + 1;
      v.pa = temp.aa, v.pb = temp.bb;
      v.type = 3;
      v.i = 2;
      map[v.aa][v.bb] = v;
      if (v.aa == c || v.bb == c) {
        write(v.aa, v.bb);
        return 0;
      }
      q.push(v);
    }
  }
  return -1;
}

int main() {
  int a, b, c;
  while (scanf("%d%d%d", &a, &b, &c) != EOF) {
    a = bfs(a, b, c);
    if (a == -1) printf("impossible\n");
  }
  return 0;
}
